src/main/ruby/core/string.rb

```diff
@@ -220,15 +220,15 @@ def scan(pattern)
       val = (match.length == 1 ? match[0] : match.captures)
       val.taint if taint
 
-      if block_given?
-        Truffle.invoke_primitive(:regexp_set_last_match, match)
+      if block
+        Truffle::RegexpOperations.set_last_match(match, block.binding)
```

Comment:
So this one was wrongly setting `$~` in the caller frame instead of the block?

Reply:
Yes, as far as I can tell.

Reply:
Can you add a spec to verify it?
It's not ideal this commit also fixes that supposed bug, as it's very hard to find the change in the middle of the rest. It would great to separate the fix in another commit if possible.

Reply:
I've pulled this change out since MRI doesn't appear to set `$~` in blocks passed to scan.